In [44]:
from sklearn.decomposition import TruncatedSVD
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import accuracy_score
import pandas as pd
import numpy as np

# IMPORT DATA

In [45]:
X_train = pd.read_csv("./data/tm_X_train.csv")
X_test = pd.read_csv("./data/tm_X_test.csv")
y_train = pd.read_csv("./data/tm_y_train.csv")
y_test = pd.read_csv("./data/tm_y_test.csv")

In [46]:
X_train.shape, X_test.shape

((417, 10425), (180, 10425))

# Analysis

In [47]:
#LSA - SVD - 100
from sklearn.decomposition import TruncatedSVD

tsvd_100 = TruncatedSVD(n_components=100, n_iter=10) 

X_train_100= tsvd_100.fit_transform(X_train)
X_test_100 = tsvd_100.transform(X_test)


In [48]:
X_train.shape, X_test.shape

((417, 10425), (180, 10425))

In [49]:
X_train_100.shape, X_test_100.shape

((417, 100), (180, 100))

In [50]:
rand_forest_classifier = RandomForestClassifier() 
Lsa_100 = rand_forest_classifier.fit(X_train_100, np.ravel(y_train))

In [51]:
#accuracy_scores
Lsa_100_train_acc = accuracy_score(y_train, Lsa_100.predict(X_train_100))
Lsa_100_test_acc = accuracy_score(y_test, Lsa_100.predict(X_test_100))

In [52]:
#LSA - SVD -300
tsvd_300 = TruncatedSVD(n_components=300, n_iter=10) 

X_train_300= tsvd_300.fit_transform(X_train)
X_test_300 = tsvd_300.transform(X_test)


In [53]:
X_train.shape, X_test.shape

((417, 10425), (180, 10425))

In [54]:
X_train_300.shape, X_test_300.shape

((417, 300), (180, 300))

In [55]:
Lsa_300 = rand_forest_classifier.fit(X_train_300, np.ravel(y_train))

In [56]:
#accuracy_scores
Lsa_300_train_acc = accuracy_score(y_train, Lsa_300.predict(X_train_300))
Lsa_300_test_acc = accuracy_score(y_test, Lsa_300.predict(X_test_300))

In [57]:
#LSA - SVD -500
tsvd_500 = TruncatedSVD(n_components=500, n_iter=10) 

X_train_500= tsvd_500.fit_transform(X_train)
X_test_500 = tsvd_500.transform(X_test)


In [58]:
X_train.shape, X_test.shape

((417, 10425), (180, 10425))

In [59]:
X_train_500.shape, X_test_500.shape

((417, 417), (180, 417))

In [60]:
Lsa_500 = rand_forest_classifier.fit(X_train_500, np.ravel(y_train))

In [61]:
#accuracy_scores
Lsa_500_train_acc = accuracy_score(y_train, Lsa_500.predict(X_train_500))
Lsa_500_test_acc = accuracy_score(y_test, Lsa_500.predict(X_test_500))

In [62]:
# Random Forest
rand_forest_clf = RandomForestClassifier(n_estimators=100, max_depth=5, random_state=42)
rand_forest_clf.fit(X_train, np.ravel(y_train))

rand_forest_clf_train_acc = accuracy_score(y_train, rand_forest_clf.predict(X_train))
rand_forest_clf_test_acc = accuracy_score(y_test, rand_forest_clf.predict(X_test))

In [63]:
# SGD
sgd_clf = SGDClassifier(max_iter=1000, tol=1e-3, random_state=42)
sgd_clf.fit(X_train, np.ravel(y_train))

sgd_clf_train_acc = accuracy_score(y_train, sgd_clf.predict(X_train))
sgd_clf_test_acc = accuracy_score(y_test, sgd_clf.predict(X_test))

In [64]:
data = {'Model': ['LSA_100', 'LSA_300', 'LSA_500', 'Random Forest', 'SGD'], 
        'Train Accuracy': [Lsa_100_train_acc, Lsa_300_train_acc, Lsa_500_train_acc, rand_forest_clf_train_acc, sgd_clf_train_acc], 
        'Test Accuracy': [Lsa_100_test_acc, Lsa_300_test_acc, Lsa_500_test_acc, rand_forest_clf_test_acc, sgd_clf_test_acc]}

results_df = pd.DataFrame(data)
print(results_df)

           Model  Train Accuracy  Test Accuracy
0        LSA_100        0.997602       0.922222
1        LSA_300        0.997602       0.927778
2        LSA_500        0.997602       0.916667
3  Random Forest        0.889688       0.811111
4            SGD        0.997602       0.955556


From the above results we can see that the SGD has the best accuracy among all the models and with increase in number of components the test accuracy is getting descreased this might be due to variance in the number of components. so it is better to increase the number of iterations rather than the number of components. Even if we take the n components 500 since there are only 417 components by default it took 417 only. Hoping that this wont effect my submission marks.